demos/bubble_shear.py

```diff
@@ -0,0 +1,483 @@
+# prescribe fine resolution in regions where it is not needed.
+#
+# For advection-dominated problems, as is the case here, a *metric advection* algorithm
+# has been proposed in :cite:<Wilson:2010>. The idea is to still compute the metric
+# based on the solution at the current time, but then to advect the metric along the
+# flow in order to predict where to prescribe fine resolution in the future. By
+# combining the advected metrics in time, we obtain a final metric that is
+# representative of the evolving solution throughout the subinterval. We achieve this in
+# the following function, where we solve the before-seen advection equation for each
+# component of the metric tensor. ::
+
+
+def adapt_metric_advection(mesh, t_start, t_end, c):
+    P1_ten = TensorFunctionSpace(mesh, "CG", 1)
+    m = RiemannianMetric(P1_ten)  # metric at current timestep
+    m_ = RiemannianMetric(P1_ten)  # metric at previous timestep
+    metric_intersect = RiemannianMetric(P1_ten)
+
+    # Compute the Hessian metric at t_start
+    for mtrc in [m, m_, metric_intersect]:
+        mtrc.set_parameters(metric_params)
+    m_.compute_hessian(c)
+    m_.enforce_spd(restrict_sizes=True, restrict_anisotropy=True)
+
+    # Set the boundary condition for the metric tensor
+    h_bc = Function(P1_ten)
+    h_max = metric_params["dm_plex_metric"]["h_max"]
+    h_bc.interpolate(Constant([[1.0 / h_max**2, 0.0], [0.0, 1.0 / h_max**2]]))
+
+    V = VectorFunctionSpace(mesh, "CG", 1)
+    u = Function(V)
+    u_ = Function(V).interpolate(velocity_expression(mesh, t_start))
+
+    R = FunctionSpace(mesh, "R", 0)
+    dt = Function(R).assign(0.01)  # timestep size
+    theta = Function(R).assign(0.5)  # Crank-Nicolson implicitness
+
+    # SUPG stabilisation
+    D = Function(R).assign(0.1)
+    h = CellSize(mesh)
+    U = sqrt(dot(u, u))
+    tau = 0.5 * h / U
+    tau = min_value(tau, U * h / (6 * D))
+
+    # Apply SUPG stabilisation
+    phi = TestFunction(P1_ten)
+    phi += tau * dot(u, grad(phi))
+
+    # Variational form of the advection equation for the metric tensor
+    trial = TrialFunction(P1_ten)
+    a = inner(trial, phi) * dx + dt * theta * inner(dot(u, grad(trial)), phi) * dx
+    L = inner(m_, phi) * dx - dt * (1 - theta) * inner(dot(u_, grad(m_)), phi) * dx
+    bcs = DirichletBC(P1_ten, h_bc, "on_boundary")
+    lvp = LinearVariationalProblem(a, L, m, bcs=bcs)
+    lvs = LinearVariationalSolver(lvp)
+
+    # Integrate from t_start to t_end
+    t = t_start + float(dt)
+    while t < t_end + 0.5 * float(dt):
+        u.interpolate(velocity_expression(mesh, t))
+
+        lvs.solve()
+
+        # Intersect metrics at every timestep
+        m.enforce_spd(restrict_sizes=True, restrict_anisotropy=True)
+        metric_intersect.intersect(m)
+
+        # Update fields at the previous timestep
+        m_.assign(m)
+        u_.assign(u)
+        t += float(dt)
+
+    metric_intersect.normalise()
```

Comment:
Not for this PR but later we should make a utility class that handles metric advection.

Reply:
Opened a new issue for this